### ISSUE TITLE
Feature request: client-side schema validation option

### ISSUE THREAD

**Issue by GabeAtWork** [ts-rest/ts-rest#306]
# Summary

This issue is a request for a new feature: allow the client to automatically run the contact's schemas on the body upon receiving a response via a new option in `initClient`.

Rationale:
- If the server/client get out of sync with a breaking change, this would help it be caught when the query runs, as early as possible
- The schemas can operate some transformation (e.g. decoding a string date), alleviating the need for the client code to write and apply their own schema

# Proposed spec

Add a new option `validateBodyWithContract` to `ClientArgs`:
```ts
export interface ClientArgs {
  baseUrl: string;
  baseHeaders: Record<string, string>;
  api?: ApiFetcher;
  credentials?: RequestCredentials;
  jsonQuery?: boolean;
  // New option here
  validateBodyWithContract?: boolean;
}
```

If the route's definition includes a schema for the response code (i.e. in my understanding, if the contract is `z.ZodTypeAny | ContractPlainType<unknown>`, which we may potentially want to simplify to the contract having a `parse` method), it gets used to parse the body in `getRouteQuery`.

# Usage

One could enable this simply by doing:

```ts
const client = initClient(contract, {
  validateBodyWithContract: true
})
```

# Alternatives

In order to solve the same problem, we could also:
* Add an option on the route definition in the contract (upside: more granular, downside: the contract now decides whether the validation should run or not)
* Add a new option `initClient` to enable/disable this behaviour on a per-route basis (upside: more granular, downside: more complex, may be kept as a future improvement)

# Misc

Here's approximately what the implementation could look like:

```ts
if (knownResponseStatuses.includes(response.status.toString())) {
      // TODO make it work with throwOnUnknownStatus disabled
      if(clientArgs.validateWithSchemas) {
        const routeResponseSchema = route.responses[response.status];
        if(!routeResponseSchema || !("parse" in routeResponseSchema)) {
          console.warn('Route schema does not have a parse method, cannot parse', {
            status: response.status
          });
          return response;
        }

        const parsedBody = routeResponseSchema.parse(response.body);

        return {...response, body: parsedBody};
      }

      return response;
    }
```

**Comment by SeanCassiere** [comment]
Got my 👀 on this one.

Was porting over an app yesterday to ts-rest when I noticed that my zod validation schema wasn't really being called since I had a transform on a field which was never applied.

Basically led to Typescript thinking a property in my response was of a certain type, which it wasn't.

**Comment by stychu** [comment]
Yeah I had the same problem recently which is unfortunate a little bit. I believe you could implement custom api handler and do all type checking there (have to check it myself) but ti would be awesome to have an ability just for the "client" version to also do the validation of the request params query response. 

@oliverbutler is it feasible for just the `client` part to have such optional functionality build in???

**Comment by SeanCassiere** [comment]
There's an issue with the types being used on the ApiFetcher type, but this fetcher does work.

You can get the zod schemas based on the response code using `args.route.responses[200]?.parse(data)`.

```ts
const customFetchApi: ApiFetcher = async ({
  path,
  method,
  headers,
  body,
  credentials,
  signal,
  cache,
  next,
  route,
}) => {
  const result = await fetch(path, {
    method,
    headers,
    body,
    credentials,
    signal,
    cache,
    next,
  } as RequestInit);
  const contentType = result.headers.get("content-type");

  if (contentType?.includes("application/") && contentType?.includes("json")) {
    const json = await result.json();

    return {
      status: result.status,
      body:
        result.status in route.responses
          ? // @ts-expect-error
            route.responses[result.status]?.parse(json)
          : json,
      headers: result.headers,
    };
  }

  if (contentType?.includes("text/plain")) {
    return {
      status: result.status,
      body: await result.text(),
      headers: result.headers,
    };
  }

  return {
    status: result.status,
    body: await result.blob(),
    headers: result.headers,
  };
};
```

**Comment by SeanCassiere** [comment]
@michaelangeloio 

Looks like the `validateResponseOnClient` option isn't being passed to the `tsRestFetchApi` caller from #373.

I think this has something to do with the `recursivelyApplyOptions` [function](https://github.com/ts-rest/ts-rest/blob/main/libs/ts-rest/core/src/lib/dsl.ts#L224-L248).

Currently, if I set the option to `true`, within the `api` caller function, it works as expected.
```typescript
const apiClient = initClient(contract, {
  baseUrl: apiBaseUrl,
  baseHeaders: {},
  api: async (args) => {
    args.route.validateResponseOnClient = true;
    return tsRestFetchApi(args);
  },
});
```
I'm a bit hesitant to touch the `recursivelyApplyOptions` function, I'm unsure of its impacts on the server aspect. 

This could be something that fixes it, but yet again I'm a bit hesitant on it.
```typescript
const recursivelyApplyOptions = <T extends AppRouter>(
  router: T,
  options?: RouterOptions
): T => {
  return Object.fromEntries(
    Object.entries(router).map(([key, value]) => {
      if (isAppRoute(value)) {
        return [
          key,
          {
            ...value,
            path: options?.pathPrefix
              ? options.pathPrefix + value.path
              : value.path,
            headers: zodMerge(options?.baseHeaders, value.headers),
            strictStatusCodes:
              value.strictStatusCodes ?? options?.strictStatusCodes,
            validateResponseOnClient: value.validateResponseOnClient ?? options?.validateResponseOnClient, // <- putting this here
          },
        ];
      } else {
        return [key, recursivelyApplyOptions(value, options)];
      }
    })
  );
};
```

**Comment by SeanCassiere** [comment]
@GabeAtWork when you have some time could you please see this [comment above](https://github.com/ts-rest/ts-rest/issues/306#issuecomment-1700242150), about the `validateResponseOnClient` flag preference not being passed down to the fetcher.

**Comment by SeanCassiere** [comment]
> @GabeAtWork when you have some time could you please see this [comment above](https://github.com/ts-rest/ts-rest/issues/306#issuecomment-1700242150), about the `validateResponseOnClient` flag preference not being passed down to the fetcher.

Will be fixed on merge and publish of #395.

**Comment by GabeAtWork** [comment]
Hey @SeanCassiere, thanks a lot for looking into it 🙏 I'll try it out as soon as I can 😊
